File: examples/nodejs/02-streaming.ts
/**
 * Streaming Example
 *
 * Demonstrates real-time streaming of completion responses.
 *
 * Requirements:
 * - Set OPENAI_API_KEY environment variable
 *
 * Run:
 *   npx ts-node 02-streaming.ts
 */

import {
    JsModelSuiteClient as ModelSuiteClient,
    JsMessage as Message,
    JsCompletionRequest as CompletionRequest,
} from 'modelsuite'

async function streamWithAsyncIterator() {
    const client = ModelSuiteClient.fromEnv()

    // Create a request with streaming enabled
    // Use "provider/model" format for explicit provider routing
    const request = CompletionRequest
        .create('openai/gpt-4o', [
            Message.user('Write a short poem about programming. 4 lines maximum.')
        ])
        .withMaxTokens(200)
        .withStreaming()

    console.log('Streaming response (async iterator):\n')

    // Stream using async iterator
    const stream = await client.stream(request)

    let chunk
    while ((chunk = await stream.next()) !== null) {
        // Print text chunks as they arrive
        if (chunk.text) {
            process.stdout.write(chunk.text)
        }

        // Check if streaming is complete
        if (chunk.isDone) {
            console.log('\n\n[Stream complete]')

            // Print final usage if available
            if (chunk.usage) {
                console.log('Total tokens:', chunk.usage.totalTokens())
            }
            break
        }
    }
}

async function streamWithCallback() {
    const client = ModelSuiteClient.fromEnv()

    const request = CompletionRequest
        .create('openai/gpt-4o', [
            Message.user('Say hello in 3 languages')
        ])
        .withMaxTokens(100)
        .withStreaming()

    console.log('\nStreaming response (callback):\n')

    // Return a promise that resolves when streaming is done
    return new Promise<void>((resolve, reject) => {
        client.completeStream(request, (chunk, error) => {
            if (error) {
                console.error('Stream error:', error)
                reject(new Error(error))
                return
            }

            if (chunk?.text) {
                process.stdout.write(chunk.text)
            }

            if (chunk?.isDone) {
                console.log('\n\n[Stream complete]')
                resolve()
            }
        })
    })
}

async function main() {
    console.log('=' .repeat(50))
    console.log('Example 1: Async Iterator')
    console.log('=' .repeat(50))
    await streamWithAsyncIterator()

    console.log('\n' + '=' .repeat(50))
    console.log('Example 2: Callback')
    console.log('=' .repeat(50))
    await streamWithCallback()
}

main().catch(console.error)

File: examples/python/02_streaming.py
"""
Streaming Example

Demonstrates real-time streaming of completion responses.

Requirements:
- Set OPENAI_API_KEY environment variable

Run:
    python 02_streaming.py
"""

from llmkit import LLMKitClient, Message, CompletionRequest


def main():
    client = LLMKitClient.from_env()

    # Create a request with streaming enabled
    # Use "provider/model" format for explicit provider routing
    request = CompletionRequest(
        model="openai/gpt-4o",
        messages=[
            Message.user("Write a short poem about programming. 4 lines maximum.")
        ],
        max_tokens=200,
    ).with_streaming()

    print("Streaming response:\n")

    # Stream the response
    for chunk in client.complete_stream(request):
        # Print text chunks as they arrive
        if chunk.text:
            print(chunk.text, end="", flush=True)

        # Check if streaming is complete
        if chunk.is_done:
            print("\n\n[Stream complete]")

            # Print final usage if available
            if chunk.usage:
                print(f"Total tokens: {chunk.usage.total_tokens()}")
            break

    print()


def stream_with_events():
    """Alternative: inspect all stream events."""
    client = LLMKitClient.from_env()

    request = CompletionRequest(
        model="openai/gpt-4o",
        messages=[Message.user("Say hello in 3 languages")],
        max_tokens=100,
    ).with_streaming()

    print("\nStreaming with event inspection:\n")

    for chunk in client.complete_stream(request):
        # Inspect event type
        event_type = chunk.event_type
        print(f"[Event: {event_type}]", end=" ")

        if chunk.text:
            print(f"Text: {chunk.text!r}")
        elif chunk.is_done:
            print(f"Stop reason: {chunk.stop_reason}")
        else:
            print()


if __name__ == "__main__":
    main()
    # Uncomment to see event-level streaming:
    # stream_with_events()
